vnext/Desktop/Modules/BlobModule.cpp

```diff
@@ -0,0 +1,239 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "BlobModule.h"
+
+#include <unicode.h>
+
+// React Native
+#include <cxxreact/Instance.h>
+#include <cxxreact/JsArgumentHelpers.h>
+#include <cxxreact/ModuleRegistry.h>
+
+// Windows API
+#include <winrt/Windows.Foundation.h>
+#include <winrt/Windows.Security.Cryptography.h>
+
+using namespace facebook::xplat;
+
+using folly::dynamic;
+using std::lock_guard;
+using std::mutex;
+using std::string;
+using std::vector;
+using std::weak_ptr;
+using winrt::Windows::Foundation::GuidHelper;
+using winrt::Windows::Security::Cryptography::CryptographicBuffer;
+
+namespace {
+constexpr char moduleName[] = "BlobModule";
+constexpr char blobURIScheme[] = "blob";
+
+weak_ptr<Microsoft::React::IWebSocketModuleContentHandler> s_contentHandler;
+} // namespace
+
+namespace Microsoft::React {
+
+BlobModule::BlobModule() noexcept {
+  m_contentHandler = std::make_shared<BlobWebSocketModuleContentHandler>();
+  s_contentHandler = m_contentHandler;
+}
+
+#pragma region CxxModule overrides
+
+string BlobModule::getName() {
+  return moduleName;
+}
+
+std::map<string, dynamic> BlobModule::getConstants() {
+  return {{"BLOB_URI_SCHEME", blobURIScheme}, {"BLOB_URI_HOST", {}}};
+}
+
+// clang-format off
+std::vector<module::CxxModule::Method> BlobModule::getMethods() {
+  return
+  {
+    Method(
+      "addNetworkingHandler",
+      [this](dynamic args)
+      {
+        //TODO: Implement
+      }
+    ),
+
+    Method(
+      "addWebSocketHandler",
+      [this](dynamic args)
+      {
+        auto id = jsArgAsInt(args, 0);
+
+        m_contentHandler->Register(id);
+      }
+    ),
+
+    Method(
+      "removeWebSocketHandler",
+      [this](dynamic args)
+      {
+        auto id = jsArgAsInt(args, 0);
+
+        m_contentHandler->Unregister(id);
+      }
+    ),
+
+    Method(
+      "sendOverSocket",
+      [this](dynamic args)
+      {
+        auto blob = jsArgAsObject(args, 0);
+        auto blobId = blob["blobId"].getString();
+        auto offset = blob["offset"].getInt();
+        auto size = blob["size"].getInt();
+        auto socketID = jsArgAsInt(args, 1);
+
+        auto data = m_contentHandler->ResolveMessage(std::move(blobId), offset, size);
+
+        if (auto instance = getInstance().lock())
+        {
+          auto buffer = CryptographicBuffer::CreateFromByteArray(data);
+          auto winrtString = CryptographicBuffer::EncodeToBase64String(std::move(buffer));
+          auto base64String = Common::Unicode::Utf16ToUtf8(std::move(winrtString));
+
+          auto sendArgs = dynamic::array(std::move(base64String), socketID);
+          instance->callJSFunction("WebSocketModule", "sendBinary", std::move(sendArgs));
```

Comment:
This feels pretty inefficient. We need to do conversion of the blob multiple times, but also `callJSFunction` will asynchronously re-dispatch this work to the JS thread, which will them jump back to the NM thread.

The Android implementation of this seems to call a function directly on the module that's overloaded to take a byte buffer to cut down on this overhead. We should look at doing something similar.

https://github.com/facebook/react-native/blob/aeaf286c77b50a95c4961de0d2355caad8ffa396/ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java

Reply:
I'll revise this workflow, but calling the module class directly is not an option because it is stored as `std::unique_ptr` deep within the `Instance` object.

Android (and iOS) solves this by adding circular reference between Blob and WS modules, having Blob own the binary data, and WS own the handler references (to the Blob module itself).

Reply:
> The Android implementation of this seems to call a function directly on the module that's overloaded to take a byte buffer to cut down on this overhead. We should look at doing something similar.
 
It does, but that's because of the circular reference between both modules.
To my knowledge, we also can't have `CxxModule` A directly call `CxxModule` B, due to their scope and exposure.

The reason Android and iOS are able to is because the module instances are exposed via a context and bridge object respectively (see PR description).